bazel/repositories.bzl

```diff
@@ -1,9 +1,54 @@
 load(":genrule_repository.bzl", "genrule_repository")
 load(":patched_http_archive.bzl", "patched_http_archive")
-load(":repository_locations.bzl", "REPO_LOCATIONS")
+load(":repository_locations.bzl", "REPOSITORY_LOCATIONS")
 load(":target_recipes.bzl", "TARGET_RECIPES")
 
-def _repository_impl(ctxt):
+def _repository_impl(name, **kwargs):
+  existing_rule_keys = native.existing_rules().keys()
+  if name in existing_rule_keys:
+    # This repository has already been defined, probably because the user
+    # wants to override the version. Do nothing.
+    return
+
+  location = REPOSITORY_LOCATIONS[name]
+
+  # Git tags are mutable. We want to depend on commit IDs instead. Give the
+  # user a useful error if they accidentally specify a tag.
+  if "tag" in location:
+    fail("Refusing to depend on Git tag %r for external dependency %r: use 'commit' instead." % (location["tag"], name))
+
+  if "commit" in location:
+    # Git repository at given commit ID. Add a BUILD file if requested.
+    if "build_file" in kwargs:
+      native.new_git_repository(
```

Comment:
Given that we can use URLs + `http_archive` to access GitHub at a given SHA, should we just kill support completely for `git_repository`? This will make things much nicer for folks who have complex internal firewall and proxy setups, where the Java git library used by Bazel has proven to be an issue in the past. Do we need `git_repository` for anything other than GH?

Reply:
Github's archive SHAs are not stable. When they change, all builds using them break.

Are the people with complex firewalls able to use normal command-line git? There's a [standard upstream Skylark rule](https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/git.bzl) that calls git directly.

Reply:
Either is fine, `http_archive` provides better performance than either `git` implementation. Also we might want to document to use `--override_repository` command line option for those behind complex firewalls.

Reply:
I think folks have usually had a command-line git workaround, so the Skylark rule may be preferable. I rather it be trivial to update the fetch rather than have folks use `--override_repository`; ideally we just interop with standard HTTP proxy or native git config, which is least surprising.

@jmillikin-stripe I read up a bit on the SHA stability issues and agree. We need to be more consistent here, since I added a GH SHA archive in `data-plane-api` based on feedback that this was a good practice.

Reply:
OK, I've changed over to the Skylark rules for Git.

@lizan is there any external you'd like to be an http_archive instead of git_repository? The unstable SHAs are tech debt, but might be justified if some particular repo is really slow to clone.